builder/dockerfile/dispatchers.go

```diff
@@ -187,28 +187,43 @@ func dispatchCopy(b *Builder, args []string, attributes map[string]bool, origina
 		return err
 	}
 
-	var contextID *int
+	var im *imageMount
 	if flFrom.IsUsed() {
 		var err error
-		context, err := strconv.Atoi(flFrom.Value)
-		if err != nil {
-			return errors.Wrap(err, "from expects an integer value corresponding to the context number")
-		}
-		if err := b.imageContexts.validate(context); err != nil {
-			return err
+		fromIndex, err := strconv.Atoi(flFrom.Value)
+		if err == nil {
+			if err := b.imageContexts.validate(fromIndex); err != nil {
+				return err
+			}
+			im = b.imageContexts.list[fromIndex]
+		} else {
+			val := strings.ToLower(flFrom.Value)
+			if m, ok := b.imageContexts.byName[val]; ok {
+				im = m
+			} else {
+				im, err = mountByRef(b, flFrom.Value)
+				if err != nil {
+					return errors.Wrapf(err, "invalid from flag value %s", flFrom.Value)
+				}
+			}
 		}
-		contextID = &context
 	}
 
-	return b.runContextCommand(args, false, false, "COPY", contextID)
+	return b.runContextCommand(args, false, false, "COPY", im)
 }
 
 // FROM imagename
 //
 // This sets the image the dockerfile will build on top of.
 //
 func from(b *Builder, args []string, attributes map[string]bool, original string) error {
-	if len(args) != 1 {
+	ctxName := ""
+	if len(args) == 3 && strings.EqualFold(args[1], "as") {
+		ctxName = strings.ToLower(args[2])
```

Comment:
Isn't it confusing when `foo` and `FOO` can mean different things? Dockerfile commands  for example are not case sensitive.